src/main/java/uk/ac/cam/cl/dtg/isaac/api/managers/EventBookingManager.java

```diff
@@ -148,21 +149,72 @@ public Long countNumberOfBookingsWithStatus(final String eventId, final BookingS
         return v;
     }
 
+    /**
+     * Create booking on behalf of a user.
+     * This method will allow users to be booked onto an event providing there is space. If there is no space the user
+     * will simply be added to the waiting list.
+     *
+     * @param event - of interest
+     * @param user  - user to book on to the event.
+     * @param additionalEventInformation - any additional information for the event organisers (nullable)
+     * @return the newly created booking.
+     * @throws SegueDatabaseException    - if an error occurs.
+     * @throws DuplicateBookingException - Duplicate booking, only unique bookings.
+     */
+    public EventBookingDTO createBookingOrAddToWaitingList(final IsaacEventPageDTO event, final RegisteredUserDTO user,
+                                         final Map<String, String> additionalEventInformation)
+            throws SegueDatabaseException, DuplicateBookingException {
+        // check if already booked
+        if (this.isUserBooked(event.getId(), user.getId())) {
+            throw new DuplicateBookingException(String.format("Unable to book onto event (%s) as user (%s) is already"
+                    + " booked on to it.", event.getId(), user.getEmail()));
+        }
+
+        EventBookingDTO booking;
+        try {
+            // Obtain an exclusive database lock to lock the booking
+            this.bookingPersistenceManager.acquireDistributedLock(event.getId());
```

Comment:
It looks like the code tries to grab the lock for this event ID both here and then, after acquiring it, in the createBooking(...) method which should result in deadlock.

Reply:
Agreed! I think it should be safe in this method to not acquire a lock as if another process fills up the event before this method calls createBooking then it will add them to the waiting list which is fine and acquires a lock separately. I will just remove the acquireLock in this method.